// gatsby-cli/lib/reporter/loggers/ink/cli.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _ink = require("ink");

var _isTty = require("../../../util/is-tty");

var _gatsbyTelemetry = require("gatsby-telemetry");

var _spinner = require("../ink/components/spinner");

var _progressBar = require("../ink/components/progress-bar");

var _messages = require("../ink/components/messages");

var _error = require("./components/error");

var _develop = _interopRequireDefault(require("../ink/components/develop"));

const showProgress = (0, _isTty.isTTY)();

class CLI extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "state", {
      hasError: false
    });
    (0, _defineProperty2.default)(this, "memoizedReactElementsForMessages", []);
  }

  componentDidCatch(error, info) {
    (0, _gatsbyTelemetry.trackBuildError)(`INK`, {
      error: {
        stack: info.componentStack,
        text: error.message,
        context: {}
      }
    });
  }

  static getDerivedStateFromError(error) {
    return {
      hasError: true,
      error
    };
  }

  render() {
    const _this$props = this.props,
          _this$props$logs = _this$props.logs,
          messages = _this$props$logs.messages,
          activities = _this$props$logs.activities,
          showStatusBar = _this$props.showStatusBar;
    const _this$state = this.state,
          hasError = _this$state.hasError,
          error = _this$state.error;

    if (hasError) {
      // You can render any custom fallback UI
      return _react.default.createElement(_ink.Box, {
        flexDirection: "row"
      }, _react.default.createElement(_messages.Message, {
        level: "ACTIVITY_FAILED",
        text: `We've encountered an error: ${error.message}`
      }));
    }
    /*
      Only operation on messages array is to push new message into it. Once
      message is there it can't change. Because of that we can do single
      transform from message object to react element and store it.
      This will avoid calling React.createElement completely for every message
      that can't change.
    */


    if (messages.length > this.memoizedReactElementsForMessages.length) {
      for (let index = this.memoizedReactElementsForMessages.length; index < messages.length; index++) {
        const msg = messages[index];
        this.memoizedReactElementsForMessages.push(msg.level === `ERROR` ? _react.default.createElement(_error.Error, {
          details: msg,
          key: index
        }) : _react.default.createElement(_messages.Message, (0, _extends2.default)({
          key: index
        }, msg)));
      }
    }

    const spinners = [];
    const progressBars = [];

    if (showProgress) {
      Object.keys(activities).forEach(activityName => {
        const activity = activities[activityName];

        if (activity.status !== `IN_PROGRESS`) {
          return;
        }

        if (activity.type === `spinner`) {
          spinners.push(activity);
        }

        if (activity.type === `progress` && activity.startTime) {
          progressBars.push(activity);
        }
      });
    }

    return _react.default.createElement(_ink.Box, {
      flexDirection: "column"
    }, _react.default.createElement(_ink.Box, {
      flexDirection: "column"
    }, _react.default.createElement(_ink.Static, null, this.memoizedReactElementsForMessages), spinners.map(activity => _react.default.createElement(_spinner.Spinner, (0, _extends2.default)({
      key: activity.id
    }, activity))), progressBars.map(activity => _react.default.createElement(_progressBar.ProgressBar, {
      key: activity.id,
      message: activity.text,
      total: activity.total,
      current: activity.current,
      startTime: activity.startTime
    }))), showStatusBar && _react.default.createElement(_develop.default, null));
  }

}

var _default = CLI;
exports.default = _default;
